crates/utils/src/export_block.rs

```diff
@@ -0,0 +1,257 @@
+    transaction::StoreTransaction,
+};
+use gw_types::{
+    bytes::Bytes,
+    offchain::ExportedBlock,
+    packed::{self, GlobalState},
+    prelude::{Builder, Entity, Pack, Reader, Unpack},
+};
+
+// pub fn export_block(store: &Store, block_number: u64) -> Result<ExportedBlock> {
+pub fn export_block(snap: &StoreReadonly, block_number: u64) -> Result<ExportedBlock> {
+    let block_hash = snap
+        .get_block_hash_by_number(block_number)?
+        .ok_or_else(|| anyhow!("block {} not found", block_number))?;
+
+    let block = snap
+        .get_block(&block_hash)?
+        .ok_or_else(|| anyhow!("block {} not found", block_number))?;
+
+    let committed_info = snap
+        .get_l2block_committed_info(&block_hash)?
+        .ok_or_else(|| anyhow!("block {} committed info not found", block_number))?;
+
+    let post_global_state = snap
+        .get_block_post_global_state(&block_hash)?
+        .ok_or_else(|| anyhow!("block {} post global state not found", block_number))?;
+
+    let deposit_requests = snap
+        .get_block_deposit_requests(&block_hash)?
+        .unwrap_or_default();
+
+    let deposit_asset_scripts = {
+        let asset_hashes = deposit_requests.iter().filter_map(|r| {
+            let h: H256 = r.sudt_script_hash().unpack();
+            if h.is_zero() {
+                None
+            } else {
+                Some(h)
+            }
+        });
+        let asset_scripts = asset_hashes.map(|h| {
+            snap.get_asset_script(&h)?.ok_or_else(|| {
+                anyhow!("block {} asset script {} not found", block_number, h.pack())
+            })
+        });
+        asset_scripts.collect::<Result<Vec<_>>>()?
+    };
+
+    let withdrawals = {
+        let reqs = block.as_reader().withdrawals();
+        let extra_reqs = reqs.iter().map(|w| {
+            let h = w.hash().into();
+            snap.get_withdrawal(&h)?
+                .ok_or_else(|| anyhow!("block {} withdrawal {} not found", block_number, h.pack()))
+        });
+        extra_reqs.collect::<Result<Vec<_>>>()?
+    };
+
+    let bad_block_hashes = get_bad_block_hashes(snap, block_number)?;
+
+    let exported_block = ExportedBlock {
+        block,
+        committed_info,
+        post_global_state,
+        deposit_requests,
+        deposit_asset_scripts,
+        withdrawals,
+        bad_block_hashes,
+    };
+
+    Ok(exported_block)
+}
+
+pub fn read_block(reader: &mut impl Read) -> Result<Option<(ExportedBlock, usize)>> {
+    let mut full_size_buf = [0u8; 4];
+    if let Err(err) = reader.read_exact(&mut full_size_buf) {
+        match err.kind() {
+            ErrorKind::UnexpectedEof if full_size_buf == [0u8; 4] => return Ok(None),
+            _ => bail!(err),
+        }
+    }
+
+    let full_size = u32::from_le_bytes(full_size_buf) as usize;
+    if full_size <= 4 {
+        bail!("block corrupted, full size {}", full_size);
+    }
+
+    let mut buf = vec![0; full_size];
```

Comment:
Consider reusing the buffer? (Move it to `ExportedBlockReader`.)

Reply:
This buffer will be moved in two places:
  - https://github.com/nervosnetwork/godwoken/pull/754/files#diff-c5d1ffc9e3eca2c0823859c004accb1afdfc0f535b10b73f33b65d3394857156R120
  - https://github.com/nervosnetwork/godwoken/pull/754/files#diff-07eff2ee69a5eac43558211b5b3efdfded30a207932dd888c54a52c2fc2f16b3R180

We can optimize later if this is a bottleneck.